### PR TITLE
BENCH: asv benchmarks for alternative array backends

### PR DESCRIPTION
**Pull request by crusaderky**
- New infrastructure for running asv on the Array API backends, which shares machinery with `conftest` and `@xp_capabilities`.
- Bew bench test for `cluster.hierarchy.is_isomorphic` (part of #22989; kept in a separate PR for obvious reasons)
- New bench test for `cluster.vq.whiten`
- Bench test for `cluster.hierarchy.linkage` now runs on alternative backends

# Known issues
`spin bench -b all --compare`, as well as bare `asv`, create an environment without the Array API optional dependencies. I did not figure out how to add them.

Without the `--compare` flag it works as expected.

# Demo
```
$ spin bench -b all -t cluster.IsIsomorphic 
[50.00%] ··· ================== ========== ==========
             --                          nobs        
             ------------------ ---------------------
                  backend          100       100000  
             ================== ========== ==========
                   numpy         168±0μs    27.8±0ms 
              array_api_strict   343±0μs    29.0±0ms 
                    cupy           n/a        n/a    
                 torch:cpu       391±0μs    28.5±0ms 
                 torch:cuda        n/a        n/a    
                 dask.array      5.06±0ms   34.5±0ms 
               jax.numpy:cpu     736±0μs    30.4±0ms 
               jax.numpy:cuda      n/a        n/a    
             ================== ========== ==========

[50.00%] ···· For parameters: 'cupy', 100
              asv: skipped: SkipNotImplemented('cupy skipped by @xp_capabilities')
              
              For parameters: 'cupy', 100000
              asv: skipped: SkipNotImplemented('cupy skipped by @xp_capabilities')
              
              For parameters: 'torch:cuda', 100
              asv: skipped: SkipNotImplemented('torch skipped by @xp_capabilities')
              
              For parameters: 'torch:cuda', 100000
              asv: skipped: SkipNotImplemented('torch skipped by @xp_capabilities')
              
              For parameters: 'jax.numpy:cuda', 100
              asv: skipped: SkipNotImplemented('jax.numpy skipped by @xp_capabilities')
              
              For parameters: 'jax.numpy:cuda', 100000
              asv: skipped: SkipNotImplemented('jax.numpy skipped by @xp_capabilities')
```
# Lessons learned
- On CUDA, do not trust the documentation about the default stream being synchronous or not. Turns out that the default stream on CuPy is asynchronous.
- Typically, CUDA backends will start demonstrating size-dependent performance far beyond the point where it is sensible to run CPU backends; see https://github.com/scipy/scipy/pull/22989#issuecomment-2889223310
- JAX and ASV have bad interaction; ASV uses a forkserver and JAX doesn't like it. I managed to work around it by not importing scipy until the `setup()` but feels like a ugly hack.

### CODE REVIEW

File: benchmarks/benchmarks/common.py [hunk 1]
```diff
@@ -10,12 +10,243 @@
...
+                    self.func(self.x, plus=False)
+
+            def time_myfunc_plus(self, backend, size):
+                self.func(self.x, plus=True)
+
+            def time_myfunc_minus(self, backend, size):
+                self.func(self.x, plus=False)
+    """
+    backends = ["numpy", "array_api_strict", "cupy", "torch:cpu", "torch:cuda",
+                "dask.array", "jax.numpy:cpu", "jax.numpy:cuda"]
```
Comment (crusaderky) [inline benchmarks/benchmarks/common.py:74]: array_api_strict matters for performance, even if it's never going to be used in production, as a proxy for arbitrary wrappers around numpy, e.g. `marray`.

File: benchmarks/benchmarks/common.py [hunk 2]
```diff
@@ -10,12 +10,243 @@
...
+        You need to call this on all arrays that your setup() function creates.
+        """
+        if self.backend == "dask.array":
+            import dask
+
+            arrays = dask.persist(*arrays)
+        elif self.backend in ("jax.numpy:cpu", "jax.numpy:cuda"):
+            import jax
+
+            jax.block_until_ready(arrays)
```
Comment (crusaderky) [inline benchmarks/benchmarks/common.py:237]: Today this is unnecessary on jax:cpu, but it's not spelled out explicitly in the documentation and it's perfectly reasonable for JAX to kick off a worker thread in the future.

File: benchmarks/benchmarks/common.py [hunk 3]
```diff
@@ -10,12 +10,243 @@
...
+
+            def wrapper(*args, **kwargs):
+                res = func(*args, **kwargs)
+                cupy.cuda.get_current_stream().synchronize()
+                return res
+
+            self.func = wrapper
+
+        else:
+            assert backend in ("numpy", "array_api_strict", "cupy")
```
Comment (crusaderky) [inline benchmarks/benchmarks/common.py:223]:
```suggestion
            assert backend in ("numpy", "array_api_strict")
```

File: benchmarks/benchmarks/common.py [hunk 4]
```diff
@@ -10,12 +10,243 @@
...
+            jax.config.update("jax_enable_x64", True)
+            try:
+                jax_device = jax.devices(device)[0]
+            except RuntimeError:
+                raise SkipNotImplemented(f"{device=} not available")
+            jax.config.update("jax_default_device", jax_device)
+
+            if capabilities["jax_jit"]:
+                func = jax.jit(func, static_argnames=static_argnames,
+                                static_argnums=static_argnums)
```
Comment (crusaderky) [inline benchmarks/benchmarks/common.py:189]:
```suggestion
                               static_argnums=static_argnums)
```

File: benchmarks/benchmarks/common.py [hunk 5]
```diff
@@ -10,12 +10,243 @@
...
+
+        class MyFunc(XPBenchmark):
+            param_names = (*XPBenchmark.param_names, "size")
+            params = (*XPBenchmark.params, [5, 10])
+
+            def setup(self, backend, size):
+                super().setup(backend, myfunc, static_argnames=("plus",))
+                x = self.xp.arange(size)
+                self.x = self.synchronize(x)
+                if self.warmup:
```
Comment (crusaderky) [inline benchmarks/benchmarks/common.py:63]: One _could_ write some automatic discovery system that applies `jax.jit` automatically to all `time_` methods - as it's safe to assume that all asv parameters end up as static jax.jit parameters.
The risk of overengineering, or encroaching in the same type of magic system that jax definitely uses, is very high though.

**Comment by crusaderky** [comment]
CI failures are unrelated

**Comment by crusaderky** [comment]
CI screenshot. Not sure if we want to change CI to run the other backends?
(that would require enabling SCIPY_ARRAY_API globally, with a potential impact on everything)

image

**Comment by rgommers** [comment]
> Not sure if we want to change CI to run the other backends?

Not yet - `asv` is quite slow, so there's a runtime issue. Also it'll be too maintenance-heavy. Once we do get to the point of wanting to enable this, we will most likely want to migrate the job from CircleCI to the Cirrus runner.

> (that would require enabling SCIPY_ARRAY_API globally, with a potential impact on everything)

Yep, can't do that here.

**Comment by crusaderky** [comment]
I've added an optional `-b` parameter to `spin bench`, to match `spin test`.
Companion PR to run in pixi: https://github.com/rgommers/pixi-dev-scipystack/pull/35

**Comment by crusaderky** [comment]
Added test for `whiten`

```
$ SCIPY_XSLOW=1 pixi r -e array-api-cuda bench-array-api -t cluster.Whiten
[50.00%] ··· cluster.Whiten.time_whiten
[50.00%] ··· ================== ========== ============ ============== ================
             --                                          shape                         
             ------------------ -------------------------------------------------------
                  backend        (10, 10)   (100, 100)   (1000, 1000)   (10000, 10000) 
             ================== ========== ============ ============== ================
                   numpy         231±0μs     262±0μs       7.10±0ms        404±0ms     
              array_api_strict   406±0μs     451±0μs       6.01±0ms        410±0ms     
                    cupy         32.5±0ms    27.7±0ms      31.8±0ms        34.3±0ms    
                 torch:cpu       1.33±0ms    1.34±0ms      11.9±0ms        942±0ms     
                 torch:cuda      83.4±0ms    86.5±0ms      83.0±0ms        85.7±0ms    
                 dask.array      5.59±0ms    5.94±0ms      13.3±0ms        503±0ms     
               jax.numpy:cpu     34.4±0μs    112±0μs       1.52±0ms        427±0ms     
               jax.numpy:cuda    83.9±0μs    91.7±0μs      201±0μs         8.40±0ms    
             ================== ========== ============ ============== ================
```

**Comment by crusaderky** [comment]
Revamped test for `linkage`

```
$ pixi r -e array-api-cuda bench-array-api -t cluster.Linkage
[50.00%] ··· cluster.Linkage.time_linkage
[50.00%] ··· ================== ====== ========== ========== ========== ========== ========== ========== ==========
             --                                                           method                                   
             ------------------------- ----------------------------------------------------------------------------
                  backend        size    single    complete   average    weighted   centroid    median      ward   
             ================== ====== ========== ========== ========== ========== ========== ========== ==========
                   numpy         2000   19.8±0ms   46.7±0ms   52.6±0ms   50.6±0ms   53.8±0ms   48.4±0ms   55.9±0ms 
              array_api_strict   2000   21.6±0ms   47.8±0ms   53.5±0ms   52.1±0ms   56.2±0ms   51.1±0ms   52.8±0ms 
                    cupy         2000     n/a        n/a        n/a        n/a        n/a        n/a        n/a    
                 torch:cpu       2000   36.0±0ms   59.0±0ms   62.0±0ms   62.6±0ms   65.2±0ms   67.3±0ms   66.5±0ms 
                 torch:cuda      2000     n/a        n/a        n/a        n/a        n/a        n/a        n/a    
                 dask.array      2000   32.8±0ms   57.3±0ms   61.2±0ms   59.9±0ms   68.7±0ms   61.8±0ms   63.1±0ms 
               jax.numpy:cpu     2000   52.2±0ms   59.9±0ms   67.6±0ms   57.9±0ms   77.5±0ms   79.6±0ms   68.1±0ms 
               jax.numpy:cuda    2000     n/a        n/a        n/a        n/a        n/a        n/a        n/a    
             ================== ====== ========== ========== ========== ========== ========== ========== ==========
```

**Comment by lucascolley** [comment]
> spin bench -b all --compare, as well as bare asv, create an environment without the Array API optional dependencies. I did not figure out how to add them.

I think we may want to wait for https://github.com/airspeed-velocity/asv/pull/1450 and https://github.com/scipy/scipy/pull/22392 before tackling that.

File: benchmarks/benchmarks/common.py [hunk 6]
```diff
@@ -10,12 +10,243 @@
...
+        self.warmup = False
+
+        if backend == "torch":
+            import torch
+
+            torch.set_default_dtype(torch.float64)
+            try:
+                torch.empty(0, device=device)
+            except RuntimeError:
+                raise SkipNotImplemented(f"{device=} not available")
```
Comment (lucascolley) [inline benchmarks/benchmarks/common.py]: This is needed for me when running in the array-api pixi env on M1 mac

```suggestion
            try:
                torch.empty(0, device=device)
            except (RuntimeError, AssertionError):
                raise SkipNotImplemented(f"{device=} not available")
```

**Review by lucascolley — APPROVED** [review]
one small change needed, otherwise I think this looks good!

**Comment by crusaderky** [comment]
I've added a benchmark for `array_namespace` and the basic pattern of xp-enabled functions:

Note the eye-watering overhead of Dask (811μs) . It's genuine.
We could reduce it a bit (559μs) by using `dask.compute(scheduler='synchronous')`, but that would slow down parallelizable workloads. FWIW the profile on the distributed scheduler will be different from both. Dask is not really designed for small data or fast round-trips; this is why all new tests have a SCIPY_XSLOW option to properly show Dask performance, with chunking and multi-threading kicking in.

```
[66.67%] ··· array_api.ArrayNamespace.time_array_namespace                                                                               ok
[66.67%] ··· ================== =============
                  backend                    
             ------------------ -------------
                   numpy         3.06±0.01μs 
              array_api_strict    18.0±0.2μs 
                    cupy         2.54±0.03μs 
                 torch:cpu       2.67±0.01μs 
                 torch:cuda      2.65±0.01μs 
                 dask.array      3.08±0.02μs 
               jax.numpy:cpu      2.57±0.2μs 
               jax.numpy:cuda    2.41±0.02μs 
             ================== =============

[83.33%] ··· array_api.ArrayNamespace.time_compat_namespace                                                                              ok
[83.33%] ··· ================== =============
                  backend                    
             ------------------ -------------
                   numpy         1.34±0.03μs 
              array_api_strict    16.9±0.2μs 
                    cupy         1.35±0.03μs 
                 torch:cpu         1.51±0μs  
                 torch:cuda      1.50±0.01μs 
                 dask.array      1.78±0.01μs 
               jax.numpy:cpu       1.44±0μs  
               jax.numpy:cuda    1.40±0.02μs 
             ================== =============

[100.00%] ··· array_api.ArrayNamespace.time_trivial_func                                                                                  ok
[100.00%] ··· ================== =============
                   backend                    
              ------------------ -------------
                    numpy         3.06±0.03μs 
               array_api_strict    18.0±0.3μs 
                     cupy         3.64±0.01μs 
                  torch:cpu       2.74±0.06μs 
                  torch:cuda      7.71±0.02μs 
                  dask.array       811±0.9μs  
                jax.numpy:cpu     3.71±0.02μs 
                jax.numpy:cuda    3.42±0.01μs 
              ================== =============
```

**Comment by crusaderky** [comment]
Test failure unrelated: https://github.com/scipy/scipy/issues/23177

**Comment by crusaderky** [comment]
@lucascolley @rgommers can we push this through the finishing line?

**Comment by lucascolley** [comment]
I'm happy with it.

**Review by rgommers — APPROVED** [review]
LGTM too, let's give this a go. Thanks @crusaderky!

One potential follow-up that could be good for usability: if an array library is missing, error out immediately rather than letting each individual benchmark show an exception about the missing library.
